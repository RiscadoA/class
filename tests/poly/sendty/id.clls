type Id {
    recvty T; recv ~T; send T; close
};;

proc id(x: Id) {
    println("id recvty");
    recvty x(U);
    println("id recv");
    recv x(y);
    println("id send");
    send x(y);
    println("id close");
    close x
};;

proc main() {
    par {
        cut {
            id(id)
            |id: ~Id|
            println("lhs sendty");
            sendty id(close);
            println("lhs send");
            send id(x. println("lhs close x"); close x);
            println("lhs recv");
            recv id(x);
            println("lhs wait id");
            wait id;
            println("lhs wait x");
            wait x;
            ()
        }
        ||
        cut {
            id(id)
            |id: ~Id|
            println("rhs sendty");
            sendty id(wait);
            println("rhs send");
            send id(x. println("rhs wait x"); wait x; ());
            println("rhs recv");
            recv id(x);
            println("rhs wait id");
            wait id;
            println("rhs close x");
            close x
        }
    }
};;
