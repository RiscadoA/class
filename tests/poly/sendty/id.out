lhs sendty
lhs send
id recvty
id recv
lhs close x
id send
id close
lhs recv
lhs wait id
lhs wait x
rhs sendty
rhs send
id recvty
id recv
id send
id close
rhs recv
rhs wait id
rhs close x
rhs wait x
